cachito/workers/tasks/npm.py

```diff
@@ -43,102 +88,128 @@ def cleanup_npm_request(request_id):
 
 
 @app.task
-def fetch_npm_source(request_id):
+def fetch_npm_source(request_id, package_configs=None):
     """
     Resolve and fetch npm dependencies for a given request.
 
+    This function uses the Python ``os.path`` library to manipulate paths, so the path to the
+    configuration files may differ in format based on the system the Cachito worker is deployed on
+    (i.e. Linux vs Windows).
+
     :param int request_id: the Cachito request ID this is for
+    :param list package_configs: the list of optional package configurations submitted by the user
     :raise CachitoError: if the task fails
     """
+    if package_configs is None:
+        package_configs = []
+
     validate_npm_config()
 
     bundle_dir = RequestBundleDir(request_id)
     log.debug("Checking if the application source uses npm")
-    for lock_file in (bundle_dir.npm_shrinkwrap_file, bundle_dir.npm_package_lock_file):
-        if lock_file.exists():
-            break
-    else:
-        raise CachitoError(
-            "The npm-shrinkwrap.json or package-lock.json file must be present for the npm "
-            "package manager"
-        )
+    subpaths = []
+    if package_configs:
+        subpaths = [os.path.normpath(c["path"]) for c in package_configs if c.get("path")]
 
-    log.debug("Ensuring there is no node_modules directory present")
-    if bundle_dir.npm_deps_dir.joinpath("node_modules").exists():
-        raise CachitoError("The node_modules directory cannot be present in the source repository")
+    if not subpaths:
+        # Default to the root of the application source
+        subpaths = [os.path.curdir]
 
-    log.debug("Ensuring that the package.json file is present")
-    if not bundle_dir.npm_package_file.exists():
-        raise CachitoError("The package.json file is not present in the source repository")
+    _verify_npm_files(bundle_dir, subpaths)
 
     log.info("Configuring Nexus for npm for the request %d", request_id)
     set_request_state(request_id, "in_progress", "Configuring Nexus for npm")
     repo_name = get_npm_proxy_repo_name(request_id)
     prepare_nexus_for_js_request(repo_name)
 
-    log.info("Fetching the npm dependencies for request %d", request_id)
-    request = set_request_state(request_id, "in_progress", "Fetching the npm dependencies")
-    try:
-        package_and_deps_info = resolve_npm(str(bundle_dir.source_dir), request)
-    except CachitoError:
-        log.exception("Failed to fetch npm dependencies for request %d", request_id)
-        raise
+    npm_config_files = []
+    downloaded_deps = set()
+    for i, subpath in enumerate(subpaths):
+        log.info("Fetching the npm dependencies for request %d in subpath %s", request_id, subpath)
+        request = set_request_state(
+            request_id,
+            "in_progress",
+            f'Fetching the npm dependencies at the "{subpath}" directory"',
+        )
+        package_source_path = str(bundle_dir.source_dir.joinpath(subpath))
```

Comment:
Shouldn't this be `app_subpath` ?

Reply:
It's less efficient but it makes the code more readable and less duplication, so I changed it to use `app_subpath`.